### PR TITLE
GEODE-5420: Protect events in HAContainer from premature modification

### PR DESCRIPTION
**Pull request by mcmellawatt**
- Updated putInProgress boolean in HAWrapper to a counter to prevent prematurely setting ClientUpdateMessage to null when events are temporarily queued during a GII or message dispatcher initialization
- decAndRemoveFromHAContainer only removes when putInProgress counter and ref count are 0
- Refactored putEventInHARegion/putConditionallyInHAContainer to prevent overwriting an existing entry in the HAContainer.  Also reduces simplifies the code and reduces duplicated logic.
- Wrote missing basic HARegionQueue unit/integration tests, and an integration test to capture setting the ClientUpdateMessage property on HAEventWrapper to null prematurely
- Added new event tracing messages at debug logging level to help track similar issues in the future

Co-authored-by: Ryan McMahon <rmcmahon@pivotal.io>
Co-authored-by: Lynn Hughes-Godfrey <lhughesgodfrey@pivotal.io>

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

- [X] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by ladyVader** [comment]
I'd still like for folks to review ... as I think the only issue we have (so far) is a test issue in IntegrationTests.

We still haven't gotten the Distributed Test or Upgrade test results yet (we may have to kick it to get these to run).

IntegrationTests failed with an OOM ...
```
org.apache.geode.internal.cache.ha.HARegionQueueIntegrationTest > verifySimultaneousPutHAEventWrapperWithRegion FAILED
    java.lang.OutOfMemoryError
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:598)
        at java.util.concurrent.ForkJoinTask.reportException(ForkJoinTask.java:677)
        at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:735)
        at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:160)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:583)
        at org.apache.geode.internal.cache.ha.HARegionQueueIntegrationTest.createAndPutHARegionQueuesSimulataneously(HARegionQueueIntegrationTest.java:579)
        at org.apache.geode.internal.cache.ha.HARegionQueueIntegrationTest.verifySimultaneousPutHAEventWrapperWithRegion(HARegionQueueIntegrationTest.java:262)
```

I suspect this is a test issue ... I'm not sure if we really meant for this high number of operations (notice that the map version of this test sets this to 1000 vs. 10000):
```
   final int numQueues = 30;
   final int numOperations = 10000;
```

In addition, we're creating a new HAEventWrapper for each message * each queue ...
In reality the CacheClientNotifier (deliverMessage) hands the same HAEventWrapper to each proxy ... so we don't need 30 HAEventWrappers for each of these 10000 operations.

We could move that up out of the loop to have this:
```
    for (int i = 0; i < numOperations; i++) {
      count.set(i);

      // CacheClientNotifier creates a single HAEventWrapper containing the ClientUpdateMessage
      // and this is enqueued for each client (CacheClientNotifier.singletonRouteClientMessage)
      ClientUpdateMessage message = new ClientUpdateMessageImpl(EnumListenerEvent.AFTER_CREATE,
          (LocalRegion) dataRegion, "key", "value".getBytes(), (byte) 0x01, null,
          new ClientProxyMembershipID(), new EventID(new byte[] {1}, 1, count.get()));

      queues.parallelStream().forEach(haRegionQueue -> {
        try {
          HAEventWrapper haEventWrapper = new HAEventWrapper(message);

          testValidationWrapperSet.add(haEventWrapper);

          haRegionQueue.put(haEventWrapper);
        } catch (InterruptedException iex) {
          throw new RuntimeException(iex);
        }
      });
    }
```

**Comment by mcmellawatt** [comment]
Made the changes as suggested by @ladyVader to avoid the OOM in the failing integration test.  Also rebased on develop.

**Comment by ladyVader** [comment]
Write new tests to ensure the HAContainer is empty:
1 - when we queue events during Message Dispatcher initialization
2 - when we queue events while servicing a GII Request for an HARegionQueue

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java [hunk 1]
```diff
@@ -1692,14 +1719,38 @@ public void initializeMessageDispatcher() throws CacheException {
       }
       Conflatable nextEvent;
       while ((nextEvent = queuedEvents.poll()) != null) {
```
Comment (ladyVader) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxy.java]: Can we refactor this draining of events (with and without synchronization) instead of duplicating code (only difference is the "with synchronization" in the log message).

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/HAEventWrapper.java [hunk 2]
```diff
@@ -87,16 +87,13 @@
...
    */
-  private transient boolean putInProgress = false;
-
-  /**
-   * A value true indicates that this instance is not used in the <code>haContainer</code>. So any
-   * changes in this instance will not be visible to the <code>haContainer</code>.
-   */
-  private transient boolean isRefFromHAContainer = false;
+  @SuppressWarnings("unused")
+  private transient volatile long putInProgressCount;
```
Comment (ladyVader) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/HAEventWrapper.java]: Should this be volatile (for this atomic value as the reference count)?

File: geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java [hunk 3]
```diff
@@ -764,24 +774,28 @@ public void endGiiQueueing() {
...
+                continue;
               }
             }
+
             basicPut(value);
+
+            // The HAEventWrapper ref count must be decremented because it was
+            // incremented when it was queued in giiQueue.
             if (value instanceof HAEventWrapper) {
-              decAndRemoveFromHAContainer((HAEventWrapper) value);
```
Comment (ladyVader) [inline geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java]: Original code did a decAndRemove here ... are we accidentally leaving in HAContainer?
Should we use checkAndRemoveFromClientMsgsRegion here?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/HAEventWrapper.java [hunk 4]
```diff
@@ -411,4 +395,15 @@ public long decAndGetReferenceCount() {
     return rcUpdater.decrementAndGet(this);
   }
 
+  public long incrementPutInProgressCounter() {
+    return putInProgressCountUpdater.incrementAndGet(this);
+  }
+
+  public long decrementPutInProgressCounter() {
```
Comment (ladyVader) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/HAEventWrapper.java]: Consider logging if < 0

File: geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java [hunk 5]
```diff
@@ -3671,23 +3739,47 @@ public Conflatable getAndRemoveFromHAContainer(Conflatable conflatable) {
...
+          long refCount = haContainerKey.decAndGetReferenceCount();
           if (logger.isDebugEnabled()) {
-            logger.debug("Removing event from {}: {}", this.region.getFullPath(),
-                wrapper.getEventId());
+            logger.debug(caller + " decremented Event ID hash code: " + haContainerKey.hashCode()
+                + "; System ID hash code: " + System.identityHashCode(haContainerKey)
+                + "; Wrapper details: " + haContainerKey);
+          }
+          if (refCount == 0L && !haContainerKey.getPutInProgress()) {
+            if (haContainerKey.getReferenceCount() == 0L) {
```
Comment (ladyVader) [inline geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java]: Is this second check on refCount necessary?

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/HAEventWrapper.java (hunk 2 above)
Comment (mcmellawatt) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/HAEventWrapper.java]: After reviewing the docs for this class, the answer is yes.  "A reflection-based utility that enables atomic updates to designated volatile long fields of designated classes".  
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicLongFieldUpdater.html

Also, note that the reason we use this over a regular AtomicLong is that it provides a slight performance benefit.

File: geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java (hunk 3 above)
Comment (mcmellawatt) [reply geode-core/src/main/java/org/apache/geode/internal/cache/ha/HARegionQueue.java]: We determined that we could leak the event so we wrote a test and changed this code to prevent that from happening by invoking CacheClientNotifier.checkAndRemoveFromClientMsgsRegion().  This will both decrement the PutInProgressCount and remove from the container if necessary.

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/HAEventWrapper.java (hunk 4 above)
Comment (mcmellawatt) [reply geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/HAEventWrapper.java]: We don't currently do this for the existing ref count and we log the ref count when debug is enabled.  Therefore, we decided this isn't necessary.

**Comment by mcmellawatt** [comment]
The distributed test failures are not related and are known to be flaky.  We implemented the requested changes.
